ext/random/randomizer.c

```diff
@@ -278,6 +278,9 @@ PHP_METHOD(Random_Randomizer, __unserialize)
 		RETURN_THROWS();
 	}
 	object_properties_load(&randomizer->std, Z_ARRVAL_P(members_zv));
+	if (EG(exception)) {
+		RETURN_THROWS();
```

Comment:
This will throw an Error instead of an Exception which might be undesirable as per https://externals.io/message/118311. Consider:
```suggestion
		zend_throw_exception(NULL, "Invalid serialization data for Random\\Randomizer object", 0);
		RETURN_THROWS();
```

Reply:
This way two exceptions were thrown 🤔

Reply:
This should automatically chain the `Error` as the `$previous` of the newly thrown `Exception`, which is what I would expect as a userland developer. I could then catch the generic unserialization exception with 8.3 and also get the underlying error as helpful context.

Reply:
@kocsismate This remark effectively applies to all `__unserialize()` handlers, not just the Randomizer. At the very least `Mt19937::__unserialize()` should also be adapted for internal consistency of ext/random.

Reply:
Thank you for the reminded, I've just added the necessary code for `Mt19937`. However, this chained exception handling is very confusing for me. The error message is as follows:
```
Fatal error: Uncaught Error: Cannot create dynamic property Random\Randomizer::$foo in ....
Stack trace:
#0 [internal function]: Random\Randomizer->__unserialize(Array)
#1 ext/random/tests/03_randomizer/gh_9186_unserialize.php(4): unserialize('O:17:"Random\\Ra...')
#2 {main}

Next Exception: Invalid serialization data for Random\Randomizer object
Stack trace:
#0 [internal function]: Random\Randomizer->__unserialize(Array)
#1 ext/random/tests/03_randomizer/gh_9186_unserialize.php(4): unserialize('O:17:"Random\\Ra...')
#2 {main}
  thrown in ext/random/tests/03_randomizer/gh_9186_unserialize.php on line 4
```

Based on the error message, I'd naturally expect that the first exception should be caught, not the next one. I know that this is not a new thing (I admit that I wasn't familiar with chained exceptions so far), but I'm sure many people will be puzzled too.

Reply:
Thanks. Let me attempt to clarify this a little more, as you said you weren't familiar with it yet.

1. Chained exceptions are not something specific to internals. Userland developers can also leverage this mechanism, by setting the `$previous` property of the `Exception`, e.g. by using the third parameter of `Exception`.
2. This mechanism is actually leveraged by userland libraries that abstract away from some underlying implementation. For example the Flysystem file system abstraction explicitly documents this: https://flysystem.thephpleague.com/docs/usage/exception-handling/#always-receive-flysystem-exceptions. The same is true for the Guzzle HTTP client which uses another library to represent HTTP messages which might throw Exceptions that need to be wrapped in order for Guzzle to not violate its contract of only emitting specific Exceptions (Guzzle will only emit the `RequestException` itself, thus needs to wrap `InvalidArgumentException` of the underlying library).

```
Fatal error: Uncaught InvalidArgumentException: Status code must be an integer value between 1xx and 5xx. in /pwd/vendor/guzzlehttp/psr7/src/Response.php:157
Stack trace:
#0 /pwd/vendor/guzzlehttp/psr7/src/Response.php(101): GuzzleHttp\Psr7\Response->assertStatusCodeRange(0)
#1 /pwd/src/Handler/StreamHandler.php(120): GuzzleHttp\Psr7\Response->__construct(0, Array, Object(GuzzleHttp\Psr7\Stream), '1.0asd', '       ')
#2 /pwd/src/Handler/StreamHandler.php(55): GuzzleHttp\Handler\StreamHandler->createResponse(Object(GuzzleHttp\Psr7\Request), Array, Object(GuzzleHttp\Psr7\Stream), NULL)
*snip*
#12 /pwd/test.php(14): GuzzleHttp\Client->send(Object(GuzzleHttp\Psr7\Request))
#13 {main}

Next GuzzleHttp\Exception\RequestException: An error was encountered while creating the response in /pwd/src/Handler/StreamHandler.php:123
Stack trace:
#0 /pwd/src/Handler/StreamHandler.php(55): GuzzleHttp\Handler\StreamHandler->createResponse(Object(GuzzleHttp\Psr7\Request), Array, Object(GuzzleHttp\Psr7\Stream), NULL)
*snip*
#10 /pwd/test.php(14): GuzzleHttp\Client->send(Object(GuzzleHttp\Psr7\Request))
#11 {main}
  thrown in /pwd/src/Handler/StreamHandler.php on line 123
```

So the mechanism of chained Exceptions may already be familiar with userland developers. 

--------

That said I agree that the default stringification used by php-src is unnatural. Exceptions are chained as a singly-linked list with the outermost Exception being the list head. The natural order would be to print the "outermost" exception first with the others following after a “caused by” like Java does: https://stackoverflow.com/a/5800718/782822. php-src's current order effectively traverses the singly-linked list in reverse order. This might be something to improve in a future version, especially when looking at #9425 which will handle Exceptions during unserialization in a generic way and thus also chain Exceptions.

Userland exception handlers should get it right, though, because the obvious way to traverse them is something like this:

```
echo $e->getMessage(), PHP_EOL;
while ($e = $e->getPrevious()) { echo "Caused by: ", $e->getMessage(), PHP_EOL; }
```

Reply:
Thank you very much for the explanation! I also agree that it should be very helpful to change the traversal order later.